### PR TITLE
Add a mode to detect the formatting mode automatically based on the filename

### PR DESCRIPTION
**Pull request by vladmos**
The mode is probably supposed to be default in the future, but not now yet.

### CODE REVIEW

**Comment by vladmos** [comment]
I'm not quite sure how I should test the code in `buildifier.go`

File: buildifier/buildifier.go [hunk 1]
```diff
@@ -253,14 +256,33 @@ var diff *differ.Differ
...
-func processFile(filename string, data []byte) {
+func processFile(filename string, data []byte, inputType string) {
 	defer func() {
 		if err := recover(); err != nil {
 			fmt.Fprintf(os.Stderr, "buildifier: %s: internal error: %v\n", filename, err)
 			exitCode = 3
 		}
 	}()
 
+	defaultMode := tables.FormattingMode
```
Comment (pmbethe09) [inline buildifier/buildifier.go]: factor out to keep this function shorter -- pass back a cleanup
func setFormattingMode(inputType string) (cleanup func())

here:
cleanup := setFormattingMode(inputType)
defer cleanup()

Comment (vladmos) [reply buildifier/buildifier.go]: Like this?

File: buildifier/buildifier.go [hunk 2]
```diff
@@ -368,6 +373,28 @@ func processFile(filename string, data []byte) {
...
+	switch inputType {
+	case "build":
+		tables.FormattingMode = tables.BuildMode
+	case "bzl":
+		tables.FormattingMode = tables.DefaultMode
+	case "auto":
+		base := filepath.Base(filename)
+		if base == "BUILD" || base == "BUILD.bazel" || base == "WORKSPACE" {
+			tables.FormattingMode = tables.BuildMode
+		} else if strings.HasSuffix(filename, ".bzl") {
```
Comment (vladmos) [inline buildifier/buildifier.go]: @laurentlb should it check for the .bzl extension or should it just use the bzl mode if the filename is not equal to `BUILD` or `WORKSPACE`?

Comment (laurentlb) [reply buildifier/buildifier.go]: you can decide

Comment (vladmos) [reply buildifier/buildifier.go]: Refactored.

**Comment by vladmos** [comment]
Added a test to verify that the buildifier binary can be launched and handles the filenames with the `--type=auto` flag correctly.

File: buildifier/integration_test.sh [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+echo -e "$INPUT" > test/BUILD
+echo -e "$INPUT" > test/test.bzl
+
+$1 --type=auto test/*
+
+echo -e "$BUILD_OUTPUT" > test/BUILD.golden
+echo -e "$BZL_OUTPUT" > test/test.bzl.golden
+
+diff test/BUILD test/BUILD.golden
+diff test/test.bzl test/test.bzl.golden
```
Comment (laurentlb) [inline buildifier/integration_test.sh]: Are you sure the test fails if the diff is incorrect?
Don't you need to return non-zero?

Comment (vladmos) [reply buildifier/integration_test.sh]: Yes, diff returns a non-zero code and the whole script returns a non-zero code in that case.
If you just edit one of the variables, `bazel test //...` will crash.

Comment (laurentlb) [reply buildifier/integration_test.sh]: But the return value of the previous diff is ignored?

Comment (laurentlb) [reply buildifier/integration_test.sh]: I suggest using `set -e` at the top (https://stackoverflow.com/a/1379904/482260)

Comment (vladmos) [reply buildifier/integration_test.sh]: You're right, added `set -e` to crash on the first error.

**Pull request merged**